File: sdks/python/src/paymentflow/resources/reporting.py
"""``client.analytics``, ``client.usage`` and ``client.request_logs`` — the three reporting
surfaces (M22.6).

One module because each is a single operation and the three are read the same way; three
classes because they are three namespaces on the client and a caller should not have to know
they share a file.

Note the two different window spellings, which are the platform's rather than this SDK's:
``analytics`` takes RFC 3339 instants and ``usage`` takes calendar dates. Usage is metered per
UTC day, so a window with a time in it would imply a precision the meter does not have.
"""

from __future__ import annotations

from typing import Optional

from .._generated.models import AnalyticsSummaryResponse, RequestLogResponse, UsageSummaryResponse
from .._generated.operations import OPERATIONS
from .._pagination import CursorPage
from .._transport import RequestOptions
from ._base import Resource, query_of

__all__ = ["Analytics", "RequestLogs", "Usage"]


class Analytics(Resource):
    def retrieve_payment_summary(
        self,
        *,
        from_: Optional[str] = None,
        to: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> AnalyticsSummaryResponse:
        """Summarizes payment activity over a window, with hourly buckets.

        ``from_`` carries the trailing underscore PEP 8 prescribes for a keyword collision;
        ``from`` is a Python keyword and cannot be a parameter name. It is sent as ``from``,
        which is what the API expects.
        """
        query = query_of(**{"from": from_, "to": to})
        result: AnalyticsSummaryResponse = self._send(
            OPERATIONS["getPaymentAnalytics"], query=query, options=options
        )
        return result


class Usage(Resource):
    def retrieve(
        self,
        *,
        from_: Optional[str] = None,
        to: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> UsageSummaryResponse:
        """Retrieves your API usage, metered per UTC day.

        The window is calendar dates (``YYYY-MM-DD``), not instants. See :class:`Analytics` for
        why ``from_`` is spelled that way.
        """
        query = query_of(**{"from": from_, "to": to})
        result: UsageSummaryResponse = self._send(OPERATIONS["getUsage"], query=query, options=options)
        return result


class RequestLogs(Resource):
    def list(
        self,
        *,
        limit: Optional[int] = None,
        starting_after: Optional[str] = None,
        created_after: Optional[str] = None,
        created_before: Optional[str] = None,
        status_code: Optional[int] = None,
        method: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> CursorPage[RequestLogResponse]:
        """Lists your API calls, most recent first.

        Each row is keyed by the ``request_id`` this SDK reports on every response, so a call
        you captured in your own logs can be looked up here directly.
        """
        query = query_of(
            limit=limit,
            starting_after=starting_after,
            created_after=created_after,
            created_before=created_before,
            status_code=status_code,
            method=method,
        )
        return self._list_cursor(OPERATIONS["listRequestLogs"], query, options)

File: sdks/node/src/resources/reporting.ts
/**
 * `client.analytics`, `client.usage` and `client.requestLogs` — the three reporting surfaces
 * (M22.3).
 *
 * One file because each is a single operation and the three are read the same way; three
 * classes because they are three namespaces on the client and a caller should not have to know
 * they share a module.
 *
 * Note the two different window spellings, which are the platform's rather than this SDK's:
 * `analytics` takes RFC 3339 instants and `usage` takes calendar dates. Usage is metered per
 * UTC day, so a window with a time in it would imply a precision the meter does not have.
 */

import { OPERATIONS } from '../generated/operations.js';
import type { AnalyticsSummaryResponse, RequestLogResponse, UsageSummaryResponse } from '../generated/models.js';
import type { RequestOptions } from '../transport.js';
import type { CursorPage } from '../pagination.js';
import { Resource } from './base.js';

/** The reporting window for `analytics.retrievePaymentSummary`, as RFC 3339 instants. */
export type AnalyticsSummaryParams = {
  readonly from?: string;
  readonly to?: string;
};

export class Analytics extends Resource {
  /** Summarizes payment activity over a window, with hourly buckets. */
  retrievePaymentSummary(
    params: AnalyticsSummaryParams = {},
    options?: RequestOptions,
  ): Promise<AnalyticsSummaryResponse> {
    return this.send<AnalyticsSummaryResponse>(OPERATIONS.getPaymentAnalytics, { query: params, options });
  }
}

/** The reporting window for `usage.retrieve`, as calendar dates (`YYYY-MM-DD`). */
export type UsageSummaryParams = {
  readonly from?: string;
  readonly to?: string;
};

export class Usage extends Resource {
  /** Retrieves your API usage, metered per UTC day. */
  retrieve(params: UsageSummaryParams = {}, options?: RequestOptions): Promise<UsageSummaryResponse> {
    return this.send<UsageSummaryResponse>(OPERATIONS.getUsage, { query: params, options });
  }
}

/** What `requestLogs.list` accepts. */
export type RequestLogListParams = {
  readonly limit?: number;
  readonly starting_after?: string;
  readonly created_after?: string;
  readonly created_before?: string;
  /** Only calls that returned this status. */
  readonly status_code?: number;
  /** Only calls made with this HTTP method. */
  readonly method?: string;
};

export class RequestLogs extends Resource {
  /**
   * Lists your API calls, most recent first.
   *
   * Each row is keyed by the `requestId` this SDK reports on every response, so a call you
   * captured in your own logs can be looked up here directly.
   */
  list(params: RequestLogListParams = {}, options?: RequestOptions): Promise<CursorPage<RequestLogResponse>> {
    return this.listCursor<RequestLogResponse>(OPERATIONS.listRequestLogs, params, options);
  }
}
